functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,164 @@
+	case functionConfig.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind):
+		if _, exist, err := functionConfig.NestedSlice(fnDeprecateField); exist || err != nil {
+			return fmt.Errorf("`additionalLabelFields` has been deprecated")
+		}
+		p.NewLabels = functionConfig.NestedStringMapOrDie("labels")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("failed to configure function: input label list cannot be empty, required valid `labels` field")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			functionConfig.GetKind(), functionConfig.GetAPIVersion(), fnConfigKind)
+	}
+	return nil
+}
+
+// set labels according to the generated common label
+func (p *LabelTransformer) setLabelsInSpecs(o *fn.KubeObject) error {
+	for _, spec := range CommonSpecs {
+		if o.IsGVK(spec.Identifier.group, spec.Identifier.version, spec.Identifier.kind) {
+			err := updateLabels(&o.SubObject, spec.Path, p.NewLabels, spec.CreateIfNotPresent)
+			p.LogResult(o, spec.Path)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Transform updates the labels in the right path using configured logic
+func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+	for _, o := range objects {
+		// this label need to set for all GKV
+		defaultPath := FieldPath{"metadata", "labels"}
+		err := updateLabels(&o.SubObject, defaultPath, p.NewLabels, true)
+		p.LogResult(o, defaultPath)
+		if err != nil {
+			return err
+		}
+		// set other common labels according to specific GKV
+		err = p.setLabelsInSpecs(o)
+		if err != nil {
+			return err
+		}
+		// handle special case with slice
+		if o.IsGVK("apps", "", "StatefulSet") {
+			for _, vctObj := range o.GetMap("spec").GetSlice("volumeClaimTemplates") {
+				err = updateLabels(vctObj, defaultPath, p.NewLabels, true)
+				p.LogResult(o, defaultPath)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Logs the result of each operation, can also modify into other logs user wants
+func (p *LabelTransformer) LogResult(o *fn.KubeObject, path []string) {
+	res, _ := json.Marshal(p.NewLabels)
+	newResult := fn.Result{
+		Message:     "set labels: " + string(res),
+		Severity:    "",
+		ResourceRef: nil,
+		Field: &fn.Field{
+			Path:          strings.Join(path, "."),
+			CurrentValue:  nil,
+			ProposedValue: nil,
+		},
+		File: &fn.File{
+			Path:  o.PathAnnotation(),
+			Index: o.IndexAnnotation(),
+		},
+		Tags: nil,
+	}
+	p.Results = append(p.Results, &newResult)
+}
+
+// the update process for each label, sort the keys to preserve sequence
+func updateLabels(o *fn.SubObject, labelPath FieldPath, newLabels map[string]string, create bool) error {
+	keys := make([]string, 0)
+	for k := range newLabels {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for i := 0; i < len(keys); i++ {
+		key := keys[i]
+		val := newLabels[key]
+		newPath := append(labelPath, key)
+		_, exist, err := o.NestedString(newPath...)
```

Comment:
Are you appending the label key to the fieldpaths?  Does it work?

Reply:
well. do you think it's possible to use `SetNestedStringMap` and `NestedStringMap` here

Reply:
append works here and pass the test cases. Using SetNestedStringMap would overwrite the labels field, but here I want to upsert the labels.

Reply:
how does the `set-labels:v0.1.5` sort the label keys for the following case:
```yaml
// function config
kind: ConfigMap
data:
    aNewKey1: val1,
    bNewKey2: val2,
    cModifyKey3: val3,
    dModifyKey4: val4,
    eNewKey5: val5,
    fNewKey6: val6 
```
```yaml
// one of the ResourceList.Items
Kind: Deployment
metadata:
   labels:
       aOtherKey1: val,
       aOtherKey2: val,
       cModifyKey3: val,
       dModifyKey4: val,
       eOtherKey5: val,
       fOtherKey6: val
```

Reply:
yeah, I'm thinking of sorting the labels *after* the update: basically consider the labels as a whole (you can do that via `NestedStringMap` --> update labels --> `SetNestedStringMap`).

Reply:
I think maps cannot be sorted, SetNestedStringMap cannot ensure the sequence. It would be more effort by setting the original one to nil, then iterate all sorted keys using SetNestedString

Reply:
Not sorting the map. I mean firstly merging the current label keys and the new label keys, and then sort the new label slice, and based on the sorted slice to update the whole labels. But I can see the downside that git will show more diff if the existing labels' order are changed.
@mengqiy Any insights what label order may be more useful to users?

Reply:
> But I can see the downside that git will show more diff if the existing labels' order are changed.

We should avoid unnecessary diff since it causes noise but not provide useful info.

Reply:
Exactly, that's what I'm concerning about. On the other hand, appending new labels to the end and updating existing labels in place can end up a non-alphabetical label order. It's unclear to me whether it could have potential problems. 

Anyway, I think current approach is fair.